### PR TITLE
Capytaine v2 Hydrostatics

### PR DESCRIPTION
**Pull request by jtgrasb**
## Description
This PR updates wecopttool to use Capytaine to compute the hydrostatics and store them as part of the bem data. This removes the hydrostatics.py script. 

## Type of PR
- [ ] Bug fix
- [X] New feature
- [ ] Documentation
- [ ] Other: (specify)

## Checklist for PR
- [X] Authors read the [contribution guidelines](https://github.com/sandialabs/WecOptTool/blob/main/.github/CONTRIBUTING.md)
- [X] The pull request is from an issue branch (not main) on *your* fork, to the [main branch in WecOptTool](https://github.com/sandialabs/WecOptTool).
- [X] The authors have given the admins edit access
- [X] All changes adhere to the style guide including PEP8, Docstrings, and Type Hints.
- [ ] Modified the documentation if applicable
- [X] Modified or added a new test
- [ ] All tests pass
- [ ] [Reference or close any relevant issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)

## Additional details
A few notes:
- Changed linear_hydrodynamics to just compute friction now.
- Added xarray import to tutorial 3 and 4 for creating inertia matrix manually
Tutorials:
- 3 is working, but slightly different because I am now feeding the density into run_bem()
- 4 is working, but slightly different because stiffness is now slightly different. I haven't figured out why yet.

### CODE REVIEW

File: wecopttool/core.py [hunk 1]
```diff
@@ -2564,3 +2505,36 @@ def time_results(fd: DataArray, time: DataArray) -> ndarray:
         out = out + \
             np.real(mag)*np.cos(w*time) - np.imag(mag)*np.sin(w*time)
     return out
+
+
+def set_properties(
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:2510]: Change name. Maybe `set_fb_centers`?

File: wecopttool/core.py [hunk 2]
```diff
@@ -2200,62 +2168,33 @@ def linear_hydrodynamics(
...
-        :python:`hydrostatic_stiffness`, or :python:`friction` are both
-        provided and included in :python:`bem_data` but have different
-        values.
     """
-    vars = {'inertia_matrix': inertia_matrix, 'friction': friction,
-            'hydrostatic_stiffness': hydrostatic_stiffness}
-
     dims = ['radiating_dof', 'influenced_dof']
-
+
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:2178]:
```suggestion

```

File: wecopttool/core.py [hunk 3]
```diff
@@ -2200,62 +2168,33 @@ def linear_hydrodynamics(
...
         element method (BEM) code Capytaine, with sign convention
-        corrected.
-    inertia_matrix
-        Inertia matrix of size :python:`(ndof, ndof)`.
-        :python:`None` if included in :python:`bem_data`.
-    hydrostatic_stiffness
-        Linear hydrostatic restoring coefficient of size
-        :python:`(nodf, ndof)`.
-        :python:`None` if included in :python:`bem_data`.
+        corrected. Also includes hydrostatics.
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:2171]: "hydrostatics"

File: wecopttool/core.py [hunk 4]
```diff
@@ -2152,12 +2116,19 @@ def run_bem(
...
-        write_info = {'hydrostatics': False,
+        write_info = {'hydrostatics': True,
                       'mesh': False,
                       'wavelength': False,
                       'wavenumber': False,
                      }
     wec_im = fb.copy(name=f"{fb.name}_immersed").keep_immersed_part()
+    wec_im = set_properties(wec_im, rho=rho)
+    if not hasattr(wec_im, 'inertia_matrix') and not hasattr(wec_im, 'hydrostatic_stiffness'):
+        wec_im.compute_hydrostatics(rho=rho)
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:2127]: is `compute_hydrostatics` equivalent to `compute_rigid_body_inertia` + `compute_hydrostatic_stiffness`? If so I would remove this `if` statement, and change the other two `elif`s to `if`.

File: wecopttool/core.py [hunk 5]
```diff
@@ -312,26 +311,16 @@ def from_bem(
...
             element method (BEM) code Capytaine, with sign convention
-            corrected.
-        inertia_matrix
-           Inertia matrix of size :python:`(ndof, ndof)`.
-           :python:`None` if included in :python:`bem_data`.
-        hydrostatic_stiffness
-            Linear hydrostatic restoring coefficient of size
-            :python:`(nodf, ndof)`.
-            :python:`None` if included in :python:`bem_data`.
+            corrected. Also includes hydrostatics.
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:323]: I would not call it "hydrostatics". I would explicitly say "inertia and hydrostatic stiffness" or something like that. Not clear that the mass/inertia is included in the term "hydrostatics". At least to me.

File: wecopttool/core.py (hunk 4 above)
Comment (jtgrasb) [reply wecopttool/core.py]: The are some extra things that compute_hydrostatics() calculates such as displaced volume, waterplane area, etc. But none of those are passed to bem_data, so I agree that they're unnecessary. Changed to `if`s

File: wecopttool/core.py [hunk 6]
```diff
@@ -2152,12 +2116,19 @@ def run_bem(
...
                       'mesh': False,
                       'wavelength': False,
                       'wavenumber': False,
                      }
     wec_im = fb.copy(name=f"{fb.name}_immersed").keep_immersed_part()
+    wec_im = set_properties(wec_im, rho=rho)
+    if not hasattr(wec_im, 'inertia_matrix') and not hasattr(wec_im, 'hydrostatic_stiffness'):
+        wec_im.compute_hydrostatics(rho=rho)
+    elif not hasattr(wec_im, 'inertia_matrix'):
+        wec_im.inertia_matrix = wec_im.compute_rigid_body_inertia(rho=rho)
```
[the remark this answers is not in the document]
Comment (jtgrasb) [reply wecopttool/core.py]: I agree but now that `compute_hydrostatics()` is removed, its just `compute_rigid_body_inertia()` and `compute_hydrostatic_stiffness()`, which don't take `g` as an input.

[the remark this answers is not in the document]
Comment (cmichelenstrofer) [reply wecopttool/core.py:2127]: right, it is not needed for inertia but it is for the hydrostatic stiffness. `compute_hydrostatic_stiffness` does take `g`

File: wecopttool/core.py [hunk 7]
```diff
@@ -312,26 +311,16 @@ def from_bem(
...
         rather than running Capytaine directly, which outputs the
         results in the correct convention. The results can be saved
-        using :py:func:`wecopttool.write_netcdf`.
-
-        In addition to the Capytaine results, if the dataset contains
-        the :python:`inertia_matrix`, :python:`hydrostatic_stiffness`,
-        or :python:`friction` these do not need to be provided
-        separately.
+        using :py:func:`wecopttool.write_netcdf`. 
+        :py:func:`wecopttool.run_bem` also computes the hydrostatics
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:315]:
```suggestion
        :py:func:`wecopttool.run_bem` also computes the inertia and 
```

File: wecopttool/core.py [hunk 8]
```diff
@@ -312,26 +311,16 @@ def from_bem(
...
+        which should be included in bem_data.
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:316]:
```suggestion
        hydrostatic stiffness which should be included in bem_data.
```

File: wecopttool/core.py [hunk 9]
```diff
@@ -2152,12 +2116,17 @@ def run_bem(
...
                       'mesh': False,
                       'wavelength': False,
                       'wavenumber': False,
                      }
     wec_im = fb.copy(name=f"{fb.name}_immersed").keep_immersed_part()
+    wec_im = set_fb_centers(wec_im, rho=rho)
+    if not hasattr(wec_im, 'inertia_matrix'):
+        wec_im.inertia_matrix = wec_im.compute_rigid_body_inertia(rho=rho)
+    if not hasattr(wec_im, 'hydrostatic_stiffness'):
+        wec_im.hydrostatic_stiffness = wec_im.compute_hydrostatic_stiffness(rho=rho)
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:2129]:
```suggestion
        wec_im.hydrostatic_stiffness = wec_im.compute_hydrostatic_stiffness(rho=rho, g=g)
```

File: wecopttool/core.py [hunk 10]
```diff
@@ -706,7 +672,6 @@ def solve(self,
             If the optimizer fails for any reason other than maximum
             number of states, i.e. for exit modes other than 0 or 9.
             See :py:mod:`scipy.optimize` for exit mode details.
-
         Examples
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:675]:
```suggestion

        Examples
```

File: wecopttool/core.py [hunk 11]
```diff
@@ -883,7 +848,6 @@ def post_process(self,
             Dynamic responses in the frequency-domain.
         results_td
             Dynamic responses in the time-domain.
-
         Examples
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:851]:
```suggestion

        Examples
```

**Review by cmichelenstrofer — APPROVED** [review]

**Pull request merged**